### DISCUSSION TITLE
Saving plot results in stl files

### DISCUSSION THREAD

**Discussion by harshamarupudi56** [pyvista/pyvista#3357] [Q&A]
```

import pyvista
from pyvista import examples
mesh = examples.load_hexbeam()
another_mesh = examples.load_uniform()
plotter = pyvista.Plotter()
actor = plotter.add_mesh(mesh, color='red')
actor = plotter.add_mesh(another_mesh, color='blue')
plotter.show()
```

In a code example like this is there a way to save the resultant blue cube? It appears vtk modules are not compatible with pv.save

**Comment by banesullivan** [comment]
The stl format is a mesh format - saving nodes and connectivity, not rendering details.

ref https://en.wikipedia.org/wiki/STL_(file_format)

Is your goal to save the mesh? or is your goal to save the scene?

To save the mesh, simply run:

```py
# STL is a triangular mesh format, so must triangulate first
another_mesh.extract_geometry().tirangulate().save('mesh.stl')
```

Plenty of other formats are available from `another_mesh.save()` as well that are better suited for the actual data type. Take a look at the docs for the `.save()` method on each PyVista data type.

To save the scene, you can use one of the several `export_*` methods on the `pyvista.Plotter`. See:

[pyvista.Plotter.export_gltf](https://docs.pyvista.org/api/plotting/_autosummary/pyvista.Plotter.export_gltf.html?highlight=export#pyvista.Plotter.export_gltf)
[pyvista.Plotter.export_html](https://docs.pyvista.org/api/plotting/_autosummary/pyvista.Plotter.export_html.html?highlight=export#pyvista.Plotter.export_html)
[pyvista.Plotter.export_obj](https://docs.pyvista.org/api/plotting/_autosummary/pyvista.Plotter.export_obj.html?highlight=export#pyvista.Plotter.export_obj)
[pyvista.Plotter.export_vrml](https://docs.pyvista.org/api/plotting/_autosummary/pyvista.Plotter.export_vrml.html?highlight=export#pyvista.Plotter.export_vrml)
[pyvista.Plotter.export_vtkjs](https://docs.pyvista.org/api/plotting/_autosummary/pyvista.Plotter.export_vtkjs.html?highlight=export#pyvista.Plotter.export_vtkjs)

**Comment by harshamarupudi56** [comment]
I wanted to save the mesh but let me ask a more specific question. I created a unique structure and would like to save the final result. Saving actor gives the error 'vtkmodules.vtkRenderingOpenGL2.vtkOpenGLActor' object has no attribute 'extract_geometry'. Saving an individual cylinder variable would save a single cylinder rather then the structure I made in pyvista. 

```
pl = pv.Plotter()
cylinder = pv.Cylinder(center = (0.6,0,0), direction = (0,0,1.0), radius = .05, height = 1)
cylinder2 = pv.Cylinder(center = (-0.55,0,0), direction = (0,0,1.0), radius = .1, height = 1)
cylinder3 = pv.Cylinder(center = (0,-0.075,0), direction = (0,0,1), radius = .1, height = 1)
cylinder4 = pv.Cylinder(center = (0,.3,0), direction = (0,0,1), radius = .075, height = 1)
actor = pl.add_mesh(cylinder)
actor = pl.add_mesh(cylinder2)
actor = pl.add_mesh(cylinder3)
actor = pl.add_mesh(cylinder4)
actor = pl.add_floor(i_resolution = 100, j_resolution=100)
actor = pl.add_floor(face = '+z',i_resolution = 100, j_resolution=100)
pl.show()
actor.extract_geometry().triangulate().save("/mesh.stl")
```

**Reply by banesullivan** [reply]
`pl.add_floor` is basically just a wrapper for `pyvista.Plane`. Try making your floors using `pyvista.Plane`, then operate on the mesh objects, not the actors.

To join them, simply use the `+` operator: `joined = cylinder + cylinder2 + ...`. Then you can save the `joined` mesh
